drivers/i2c/i2c_xilinx.c

```diff
@@ -0,0 +1,641 @@
+}
+
+static int i2c_xilinx_read_dyn(const struct i2c_xilinx_config *config, struct i2c_xilinx_data *data,
+			       struct i2c_msg *msg, uint16_t addr)
+{
+	uint8_t *read_ptr = msg->buf;
+	uint32_t bytes_left = msg->len;
+	uint32_t bytes_to_read = bytes_left;
+	uint32_t cr = CR_EN;
+	uint32_t len_word = bytes_left;
+
+	if (!bytes_left || bytes_left > MAX_DYNAMIC_READ_LEN) {
+		return -EINVAL;
+	}
+	if (msg->flags & I2C_MSG_RESTART) {
+		cr |= CR_MSMS | CR_RSTA;
+	}
+	sys_write32(cr, config->base + REG_CR);
+
+	if (bytes_to_read > FIFO_SIZE) {
+		bytes_to_read = FIFO_SIZE;
+	}
+	sys_write32(bytes_to_read - 1, config->base + REG_RX_FIFO_PIRQ);
+	sys_write32((addr << 1) | I2C_MSG_READ | TX_FIFO_START, config->base + REG_TX_FIFO);
+
+	if (msg->flags & I2C_MSG_STOP) {
+		len_word |= TX_FIFO_STOP;
+	}
+	sys_write32(len_word, config->base + REG_TX_FIFO);
+
+	while (bytes_left) {
+		int ret;
+
+		bytes_to_read = bytes_left;
+		if (bytes_to_read > FIFO_SIZE) {
+			bytes_to_read = FIFO_SIZE;
+		}
+
+		sys_write32(bytes_to_read - 1, config->base + REG_RX_FIFO_PIRQ);
+		ret = i2c_xilinx_wait_rx_full(config, data, bytes_to_read);
+		if (ret) {
+			return ret;
+		}
+
+		while (bytes_to_read) {
+			*read_ptr++ = sys_read32(config->base + REG_RX_FIFO) & RX_FIFO_DATA_MASK;
+			bytes_to_read--;
+			bytes_left--;
+		}
+	}
+	return 0;
+}
+
+static int i2c_xilinx_wait_tx_done(const struct i2c_xilinx_config *config,
+				   struct i2c_xilinx_data *data, bool is_stop)
+{
+	const uint32_t finish_bit = is_stop ? ISR_BUS_NOT_BUSY : ISR_TX_FIFO_EMPTY;
+
+	uint32_t events = i2c_xilinx_wait_interrupt(
+		config, data, finish_bit | ISR_TX_ERR_TARGET_COMP | ISR_ARB_LOST);
+	if (events != finish_bit) {
+		if (!events) {
+			return -ETIMEDOUT;
+		}
+		if (events & ISR_ARB_LOST) {
+			LOG_ERR("Arbitration lost on TX");
+			return -EAGAIN;
+		}
+		LOG_ERR("TX received NAK");
+		return -ENXIO;
+	}
+	return 0;
+}
+
+static int i2c_xilinx_wait_not_busy(const struct i2c_xilinx_config *config,
+				    struct i2c_xilinx_data *data)
+{
+	if (sys_read32(config->base + REG_SR) & SR_BB) {
+		uint32_t events = i2c_xilinx_wait_interrupt(config, data, ISR_BUS_NOT_BUSY);
+
+		if (events != ISR_BUS_NOT_BUSY) {
+			LOG_ERR("Bus stuck busy");
+			i2c_xilinx_reinit(config);
+			return -EBUSY;
+		}
+	}
+	return 0;
+}
+
+static int i2c_xilinx_write(const struct i2c_xilinx_config *config, struct i2c_xilinx_data *data,
+			    const struct i2c_msg *msg, uint16_t addr)
+{
+	const uint8_t *write_ptr = msg->buf;
+	uint32_t bytes_left = msg->len;
+	uint32_t cr = CR_EN | CR_TX;
+	uint32_t fifo_space = FIFO_SIZE - 1; /* account for address being written */
+
+	if (msg->flags & I2C_MSG_RESTART) {
+		cr |= CR_MSMS | CR_RSTA;
+	}
+
+	i2c_xilinx_clear_interrupt(config, data, ISR_TX_ERR_TARGET_COMP | ISR_ARB_LOST);
+
+	sys_write32(cr, config->base + REG_CR);
+	sys_write32((addr << 1) | TX_FIFO_START, config->base + REG_TX_FIFO);
+
+	/* TX FIFO empty detection is somewhat fragile because the status register
```

Comment:
This I2C IP is a bit of a travesty in terms of not working in a nice way for a polling loop, but also not a nice way for interrupt driven transfers. Is this true for all versions of the IP?

Reply:
AFAIK, yes. Both the TX FIFO empty detection and the RX FIFO full detection (in non-dynamic mode) have some annoying quirks. The fact that the driver can inadvertently set bits in the interrupt status register by writing 1 to them when they're not actually set is another annoyance I spent a bit of time figuring out. Most of these quirks are actually documented by Xilinx in some form at least..

This IP has been around quite a while at this point, so it seems unlikely to get a change for any of these if it might break driver compatibility. I think the only change they made in the 2.1 revision was to fix a bug where dynamic read mode sometimes didn't actually wait for the read byte count to be written to the FIFO before starting the transfer.